### PR TITLE
ARROW-12983: [C++][Python][R] Properly overflow to chunked array in Python-to-Arrow conversion

### PR DESCRIPTION
**Pull request by lidavidm**
This fixes an error from when this was last refactored/improved. Now, when building a chunked array, we properly build multiple chunks. 

This also fixes a case Weston pointed out, such that an array of 2**31 strings (which won't fit in the offsets buffer) will also get chunked instead of just saying OOM.

Unfortunately the test is rather slow so I've marked it such that it doesn't run in CI.

I've updated R but note that R never builds a chunked array hence there should be no effect.

### CODE REVIEW

**Comment by nealrichardson** [comment]
> I've updated R but note that R never builds a chunked array hence there should be no effect.

Not yet! But we hope to in ARROW-9293. cc @romainfrancois

File: cpp/src/arrow/python/python_to_arrow.cc [hunk 1]
```diff
@@ -536,6 +539,41 @@ class PyPrimitiveConverter<T, enable_if_binary<T>>
     return Status::OK();
   }
 
+  Result<int64_t> ExtendAsMuchAsPossible(PyObject* values, int64_t size,
+                                         int64_t offset) override {
+    DCHECK_GE(size, offset);
+    // See BaseBinaryBuilder::Resize - avoid error in Reserve()
+    ARROW_LOG(WARNING) << "size=" << size << " offset=" << offset
```
Comment (westonpace) [inline cpp/src/arrow/python/python_to_arrow.cc:546]: Did you mean to check these log statements in?

**Review by westonpace — COMMENTED** [review]
Looks good.  Though I see don't think R is using the chunker.  I wonder if R can handle very large string arrays. 

Edit: Ah, I see you beat me to the observation :)

File: cpp/src/arrow/python/python_to_arrow.cc (hunk 1 above)
Comment (lidavidm) [reply cpp/src/arrow/python/python_to_arrow.cc]: Whoops, didn't mean to reveal my amazing debugging secrets :) (thanks for catching this)

**Comment by lidavidm** [comment]
@pitrou would you like to check over the Python changes here?

File: cpp/src/arrow/python/iterators.h [hunk 2]
```diff
@@ -84,18 +84,29 @@ inline Status VisitSequenceGeneric(PyObject* obj, VisitorFunc&& func) {
   return Status::OK();
 }
 
-// Visit sequence with no null mask
+// Visit sequence with no null mask or offset
 template <class VisitorFunc>
 inline Status VisitSequence(PyObject* obj, VisitorFunc&& func) {
```
Comment (pitrou) [inline cpp/src/arrow/python/iterators.h:89]: How about always passing `offset`? That will simplify the overloads a bit (this should not be considered a public API).

File: cpp/src/arrow/python/python_to_arrow.cc [hunk 3]
```diff
@@ -585,6 +619,37 @@ class PyPrimitiveConverter<T, enable_if_string_like<T>>
     return Status::OK();
   }
 
+  Result<int64_t> ExtendAsMuchAsPossible(PyObject* values, int64_t size,
```
Comment (pitrou) [inline cpp/src/arrow/python/python_to_arrow.cc:622]: Can you try to reconcile the `PyPrimitiveConverter` for binary and string types? They really look very similar.

**Review by pitrou — COMMENTED** [review]
I'm curious: wasn't the conversion code from @kszucs already supposed to handle this?

**Comment by kszucs** [comment]
> I'm curious: wasn't the conversion code from @kszucs already supposed to handle this?

It was. [That PR has landed in 2.0](https://github.com/apache/arrow/commit/87dd7e9894b2beb6310e634764177ab83eac3009) and according to the issue this regression appeared in version 4.0. I assume that another patch has introduced the issue again.

If I recall correctly we regularly exercised the big memory tests on buildbot since we didn't have the necessary resources available on hosted CI services. We didn't catch the regression since we decommissioned the buildbot machines.

I'm going to investigate further.

**Comment by kszucs** [comment]
I suspect that [this commit](https://github.com/apache/arrow/commit/7184c3f46981dd52c3c521b2676796e82f17da77) has introduced the regressions since this is the only one applied since version 3.0.

**Comment by lidavidm** [comment]
Ah sorry @kszucs I should've linked the commit somewhere, I also think it was that commit that introduced this. 

I've cleaned up the VisitSequence overloads and consolidated the binary/string converters.

**Comment by kszucs** [comment]
Confirmed. I executed the large memory tests for [the suspected commit](https://github.com/apache/arrow/commit/7184c3f46981dd52c3c521b2676796e82f17da77) and I instantly get a segfault.

**Comment by kszucs** [comment]
@davidli the `list` tests cases are failing for me locally. BTW couldn't we use only the `ExtendAsMuchAsPossible` methods by passing zero offset by default?

**Comment by lidavidm** [comment]
Hmm, thanks, I'll take a look.

We could do that too. That'll clean things up.

**Comment by lidavidm** [comment]
Argh, it's because I only looked at the string/binary case and not the list case, thanks for pointing that out. (Is there a crossbow build that runs these tests?)

**Comment by kszucs** [comment]
> Is there a crossbow build that runs these tests?

No, we don't have the necessary resources available on any of the CI providers.

**Comment by lidavidm** [comment]
Hmm, I'm not sure if ExtendAsMuchAsPossible should replace Extend everywhere. For one, if you use Extend, you'll then have to check that the number of items added is equal to the number of items you gave it, and in that case, you'll also lose the actual error (or we'll have to declare `Status ExtendAsMuchAsPossible(..., int64_t* appended)`). 

Maybe what we could do instead is expose a `int64_t MaxItemsPerChunk(int64_t)` so that the chunked converter can handle all the logic internally, and the converters themselves could assume everything fits in a chunk. Though that's much less flexible, it's basically the current situation.

**Comment by kszucs** [comment]
I think we need to return with a tuple/pair: `Result<(status, num_converted)>`

**Comment by kszucs** [comment]
Closing in favor of #10556

**Pull request closed without merging**